modules/apps/headless/headless-admin-workflow/headless-admin-workflow-impl/src/main/java/com/liferay/headless/admin/workflow/internal/resource/v1_0/WorkflowDefinitionLinkResourceImpl.java

```diff
@@ -133,24 +138,87 @@ public WorkflowDefinitionLink postWorkflowDefinitionWorkflowDefinitionLink(
 				workflowDefinition.getName(), workflowDefinition.getVersion()));
 	}
 
-	private WorkflowDefinitionLink _toWorkflowDefinitionLink(
-			com.liferay.portal.kernel.model.WorkflowDefinitionLink
-				workflowDefinitionLink)
+	@Override
+	public WorkflowDefinitionLink
+			putWorkflowDefinitionLinkByExternalReferenceCode(
+				String externalReferenceCode,
+				WorkflowDefinitionLink workflowDefinitionLink)
 		throws Exception {
 
-		return new WorkflowDefinitionLink() {
-			{
-				setClassName(workflowDefinitionLink::getClassName);
-				setGroupId(workflowDefinitionLink::getGroupId);
-				setId(workflowDefinitionLink::getWorkflowDefinitionLinkId);
-				setWorkflowDefinitionName(
-					workflowDefinitionLink::getWorkflowDefinitionName);
-				setWorkflowDefinitionVersion(
-					workflowDefinitionLink::getWorkflowDefinitionVersion);
-			}
-		};
+		long groupId = workflowDefinitionLink.getGroupId();
+
+		if (groupId != 0) {
+			Group group = _groupLocalService.getGroupByExternalReferenceCode(
+				workflowDefinitionLink.getGroupExternalReferenceCode(),
+				contextCompany.getCompanyId());
+
+			groupId = group.getGroupId();
+		}
+
+		com.liferay.portal.kernel.model.WorkflowDefinitionLink
+			serviceBuilderWorkflowDefinitionLink =
+				_workflowDefinitionLinkService.
+					fetchWorkflowDefinitionLinkByExternalReferenceCode(
+						externalReferenceCode, groupId);
+
+		if (serviceBuilderWorkflowDefinitionLink != null) {
```

Comment:
I think you should delegate the UPSERT logic to the service layer. See this [example](https://github.com/liferay/liferay-portal/blob/99fd09b42abaf5444cb2aaea14ecf6cb2b44282e/modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectActionResourceImpl.java#L171-L188).